src/lib.rs

```diff
@@ -1111,24 +1119,40 @@ pub unsafe trait TryFromBytes {
     /// ```
     #[must_use = "has no side effects"]
     #[inline]
-    fn try_mut_from(bytes: &mut [u8]) -> Option<&mut Self>
+    fn try_mut_from(bytes: &mut [u8]) -> Result<&mut Self, TryCastError<&mut [u8], Self>>
     where
         Self: KnownLayout + Immutable, // TODO(#251): Remove the `Immutable` bound.
     {
         util::assert_dst_is_not_zst::<Self>();
-        let candidate = Ptr::from_mut(bytes).try_cast_into_no_leftover::<Self>()?;
-
-        // This call may panic. If that happens, it doesn't cause any soundness
-        // issues, as we have not generated any invalid state which we need to
-        // fix before returning.
-        //
-        // Note that one panic or post-monomorphization error condition is
-        // calling `try_into_valid` (and thus `is_bit_valid`) with a shared
-        // pointer when `Self: !Immutable`. Since `Self: Immutable`, this panic
-        // condition will not happen.
-        let candidate = candidate.try_into_valid();
-
-        candidate.map(Ptr::as_mut)
+        let size = bytes.len();
+        match Ptr::from_mut(bytes).try_cast_into_no_leftover::<Self>() {
+            Ok(candidate) => {
+                // This call may panic. If that happens, it doesn't cause any soundness
+                // issues, as we have not generated any invalid state which we need to
+                // fix before returning.
+                //
+                // Note that one panic or post-monomorphization error condition is
+                // calling `try_into_valid` (and thus `is_bit_valid`) with a shared
+                // pointer when `Self: !Immutable`. Since `Self: Immutable`, this panic
+                // condition will not happen.
+                match candidate.try_into_valid() {
+                    Ok(candidate) => Ok(candidate.as_mut()),
+                    Err(e) => Err(e
+                        .map_src(|src| {
+                            // SAFETY: The provided `size`, below, is computed from
+                            // `bytes.len()`. By contract on `try_into_valid` we can
+                            // count on `src` being equal to `candidate`. Since
+                            // `candidate` was derived from the entire extent of
+                            // `bytes`, the size of `bytes` is exactly equal to the
+                            // size of `src`.
+                            let bytes = unsafe { src.as_bytes_with_size(size) };
+                            bytes.as_mut()
+                        })
```

Comment:
It feels sketchy to me that we have to make this implementation so much messier (esp with introducing new `unsafe`) than it was before just to support this change to our API. I wonder if there's something we can do to refactor this so it stays reasonable?

Reply:
Perhaps; I'll poke at this a bit more. This is all done to work around our `as_bytes` method having a `Sized` bound. We'll be able to remove that bound once `std::mem::size_of_val_raw` is stabilized, but until then, the workaround is to stash the size while we have it. I don't think it's likely we'll be able to avoid the new `unsafe` here.